gprofiler/metadata/py_module_version.py

```diff
@@ -0,0 +1,266 @@
+    if root is None or info is None:
+        return paths
+    try:
+        info_rel = pathlib.Path(info).relative_to(root)
+    except ValueError:  # info is not relative to root.
+        return paths
+    if not info_rel.parts:  # info *is* root.
+        return paths
+    return (_convert_legacy_entry(pathlib.Path(p).parts, info_rel.parts) for p in paths)
+
+
+def _get_package_name(dist: pkg_resources.Distribution) -> Optional[str]:
+    """Based on pip._internal.metadata.base.BaseDistribution.raw_name"""
+    metadata = _get_metadata(dist)
+    if metadata:
+        # The metadata should NEVER be missing the Name: key, but if it somehow
+        # does, fall back to the known canonical name.
+        return metadata.get("Name", dist.project_name)
+    return None
+
+
+def _get_libpython_path(pid: int) -> Optional[str]:
+    # Match the path to libpython as it may appear in /proc/[pid]/maps, e.g. " /usr/local/lib/libpython3.9.so.1.0"
+    libpython_maps_pattern = re.compile(r"(?<=\s)/\S*libpython\S*\.so(\.\S+)?\Z")
+
+    try:
+        f = open(f"/proc/{pid}/maps")
+    except OSError:
+        return None
+
+    with f:
+        for line in f.readlines():
+            match = libpython_maps_pattern.search(line.strip())
+            if match is not None:
+                return match.group()
+    return None
+
+
+@functools.lru_cache(maxsize=128)
+def _get_python_full_version(process: Process, short_version: str) -> Optional[str]:
+    assert re.match(r"[23]\.\d\d?", short_version)
+
+    bin_file = _get_libpython_path(process.pid) or f"/proc/{process.pid}/exe"
+    full_version_string_pattern = re.compile(rb"(?<=\D)" + short_version.encode() + rb"\.\d\d?(?=\x00)")
+
+    # Try to extract the version string from the binary
+    try:
+        f = open(resolve_host_path(process, bin_file), "rb")
+    except OSError:
+        return None
+
+    with f:
+        for line in f.readlines():
+            match = full_version_string_pattern.search(line)
+            if match is not None:
+                return match.group().decode()
+    return None
+
+
+def _get_standard_libs_version(result: Dict[str, Optional[Tuple[str, str]]], process: Process):
+    # Standard library modules are identified by being under a pythonx.y dir and *not* under site/dist-packages
+    standard_lib_pattern = re.compile(r"/python(?P<version>\d\.\d\d?)/(?!.*(site|dist)-packages)")
+    py_version = None
+
+    for path in result:
+        match = standard_lib_pattern.search(path)
+        if match is not None:
+            if py_version is None:
+                py_version = _get_python_full_version(process, match.group("version"))
+                if py_version is None:
+                    # No need to continue trying if we failed
+                    return None
+            result[path] = ("standard-library", py_version)  # type: ignore
+
+
+@functools.lru_cache(maxsize=128)
+def _get_dists_files(packages_path: str) -> Dict[str, pkg_resources.Distribution]:
```

Comment:
Why aren't we caching against the `Process` object here?

Reply:
Mistake, should be `Process` + `packages_path` in the process' namespace